### PR TITLE
Components are added/removed immediately

### PR DESCRIPTION
**Pull request by saltares**
* Listeners are notified after system update.
* Fixes issue #186
* Unit tests are passing

### CODE REVIEW

File: ashley/src/com/badlogic/ashley/core/Engine.java [hunk 1]
```diff
@@ -432,8 +432,12 @@ private void processComponentOperations() {
 			ComponentOperation operation = componentOperations.get(i);
 
 			switch(operation.type) {
-				case Add: operation.entity.addInternal(operation.component); break;
```
Comment (saltares) [inline ashley/src/com/badlogic/ashley/core/Engine.java]: Components are immediately added/removed. The only thing left to do (asynchronously) is listener notification and family membership update.

File: ashley/src/com/badlogic/ashley/core/Engine.java [hunk 2]
```diff
@@ -494,27 +498,20 @@ public void remove(Entity entity, Class<? extends Component> componentClass) {
 
 		public Type type;
 		public Entity entity;
-		public Component component;
```
Comment (saltares) [inline ashley/src/com/badlogic/ashley/core/Engine.java]: These are no longer needed.

File: ashley/src/com/badlogic/ashley/core/Entity.java [hunk 3]
```diff
@@ -62,11 +62,15 @@ public Entity () {
 	 * @return The Entity for easy chaining
 	 */
 	public Entity add (Component component) {
-		if (componentOperationHandler != null) {
-			componentOperationHandler.add(this, component);
-		} else {
-			addInternal(component);
+		if (addInternal(component)) {
```
Comment (saltares) [inline ashley/src/com/badlogic/ashley/core/Entity.java]: Naturally, only schedule operations or notify listeners if the component was actually added/removed.

**Comment by azakhary** [comment]
All seems legit, nothing crazy so far :D

**Comment by saltares** [comment]
Next step would be to start looking at #178 although that won't change the API.

**Pull request merged**